File: mcp/tests/opplan_ask_test.rs
//! Contract §11 interactive replies (`ask`) and the §7 auto-continuation rule.

mod common;

use stencil_mcp::opplan::{
    format_ask, loads_without_tracing, parse_op_plan, OpPlan, MAX_ASK_OPTIONS,
};

use common::plan_of;
// ── §11 interactive replies (`ask`) ──
// This server is a TOOL, not a chat: the card is validated exactly like every other client,
// then surfaced to the CALLING agent as text + structured data. Previews are never shown.

fn ask_plan(ask: &str) -> OpPlan {
    parse_op_plan(&format!(r#"{{"version":1,"reply":"pick","ask":{ask}}}"#)).unwrap()
}

fn ask_rejects(ask: &str) {
    let text = format!(r#"{{"version":1,"reply":"pick","ask":{ask}}}"#);
    assert!(
        parse_op_plan(&text).is_err(),
        "expected this card to reject the plan: {ask}"
    );
}

#[test]
fn ask_card_parses_with_its_defaults() {
    let plan = ask_plan(r#"{"question":"Which tint?","options":[{"label":"Sepia"},{"label":"B&W"}]}"#);
    let card = plan.ask.expect("card");
    assert_eq!(card.question, "Which tint?");
    assert!(!card.multi);
    assert!(!card.allow_custom);
    assert_eq!(
        card.options.iter().map(|o| o.label.as_str()).collect::<Vec<_>>(),
        vec!["Sepia", "B&W"]
    );
}

#[test]
fn ask_multi_custom_row_and_trimming() {
    let plan = ask_plan(
        r#"{"question":"  Which?  ","mode":"multi","allowCustom":true,"customLabel":" Other ","options":[{"label":"  A  "},{"label":"B"}]}"#,
    );
    let card = plan.ask.expect("card");
    assert!(card.multi);
    assert!(card.allow_custom);
    assert_eq!(card.question, "Which?");
    assert_eq!(card.custom_label, "Other");
    assert_eq!(card.options[0].label, "A");
}

#[test]
fn no_card_on_an_ordinary_or_chat_only_turn() {
    assert!(plan_of(r#"{"version":1,"reply":"hi","actions":[]}"#).ask.is_none());
    assert!(plan_of("just chatting").ask.is_none());
}

#[test]
fn previews_are_dropped_with_one_note_never_the_option() {
    let plan = ask_plan(
        r#"{"question":"Which?","options":[{"label":"Sepia","actions":[{"op":"filter","mode":"sepia"}]},{"label":"Web","image":{"url":"https://e/x.png"}}]}"#,
    );
    let card = plan.ask.expect("card");
    assert_eq!(card.options.len(), 2, "both options survive");
    let notes = plan
        .warnings
        .iter()
        .filter(|w| w.contains("option previews"))
        .count();
    assert_eq!(notes, 1, "one note per CARD, not per option");
}

#[test]
fn ask_option_action_previews_and_all_image_reference_forms_are_accepted() {
    // §11.1: options may carry `actions` (a render spec) or `image` — one of
    // url / projectId / scanIndex. This server shows no pictures, so every form
    // is accepted and only the label survives (§11.4) — never a rejection.
    let plan = ask_plan(
        r#"{"question":"Which?","options":[
            {"label":"Crop","actions":[{"op":"crop","spec":{"x1":"10%"}}]},
            {"label":"Web","image":{"url":"https://e/x.png"}},
            {"label":"Draft","image":{"projectId":"p_12"}},
            {"label":"Scan","image":{"scanIndex":3}}
        ]}"#,
    );
    let card = plan.ask.expect("card");
    assert_eq!(
        card.options.iter().map(|o| o.label.as_str()).collect::<Vec<_>>(),
        ["Crop", "Web", "Draft", "Scan"]
    );
}

#[test]
fn malformed_cards_reject_the_whole_plan() {
    ask_rejects(r#""hello""#);
    ask_rejects(r#"{"options":[{"label":"A"},{"label":"B"}]}"#);
    ask_rejects(r#"{"question":"   ","options":[{"label":"A"},{"label":"B"}]}"#);
    ask_rejects(r#"{"question":"Q","options":[]}"#);
    ask_rejects(r#"{"question":"Q","options":[{"label":"only"}]}"#);
    ask_rejects(
        r#"{"question":"Q","options":[{"label":"1"},{"label":"2"},{"label":"3"},{"label":"4"},{"label":"5"},{"label":"6"}]}"#,
    );
    ask_rejects(r#"{"question":"Q","mode":"maybe","options":[{"label":"A"},{"label":"B"}]}"#);
    ask_rejects(r#"{"question":"Q","options":[{"label":"A"},{"label":"B"}],"sneaky":1}"#);
    ask_rejects(r#"{"question":"Q","options":[{"label":"A","sneaky":1},{"label":"B"}]}"#);
    ask_rejects(r#"{"question":"Q","options":[{"label":""},{"label":"B"}]}"#);
    ask_rejects(
        r#"{"question":"Q","options":[{"label":"A","actions":[],"image":{"url":"https://e/x"}},{"label":"B"}]}"#,
    );
    ask_rejects(r#"{"question":"Q","options":[{"label":"A"},{"label":"B"}],"allowCustom":"yes"}"#);
}

#[test]
fn five_options_is_the_cap_and_it_parses() {
    let plan = ask_plan(
        r#"{"question":"Q","options":[{"label":"1"},{"label":"2"},{"label":"3"},{"label":"4"},{"label":"5"}]}"#,
    );
    assert_eq!(plan.ask.expect("card").options.len(), MAX_ASK_OPTIONS);
}

#[test]
fn a_card_is_formatted_as_a_numbered_list_for_the_calling_agent() {
    let plan = ask_plan(
        r#"{"question":"Which tint?","allowCustom":true,"options":[{"label":"Sepia"},{"label":"B&W"}]}"#,
    );
    let text = format_ask(&plan.ask.expect("card"));
    assert!(text.contains("Which tint?"), "got: {text}");
    assert!(text.contains("1. Sepia"), "got: {text}");
    assert!(text.contains("2. B&W"), "got: {text}");
    assert!(text.contains("answer freely"), "got: {text}");
    assert!(text.contains("next prompt"), "the agent is told how to answer: {text}");
}

#[test]
fn a_plan_may_both_edit_and_ask() {
    let plan = parse_op_plan(
        r#"{"version":1,"reply":"cropped; now pick","actions":[{"op":"rotate","dir":"left"}],"ask":{"question":"Which tint?","options":[{"label":"Sepia"},{"label":"B&W"}]}}"#,
    )
    .unwrap();
    assert_eq!(plan.actions.len(), 1);
    assert_eq!(plan.ask.expect("card").options.len(), 2);
}

// ── §7 auto-continuation ──

/// §7 auto-continuation: a plan whose actions load a picture (`blank`/`frame`) and drew
/// no layout LINE continues; drawn lines, variants, or an `ask` finish the turn.
#[test]
fn loads_without_tracing_follows_the_section7_rule() {
    let plan = |json: &str| parse_op_plan(json).unwrap();

    // A bare load continues; so does load + pixel-independent edits (crop/filter/page).
    assert!(loads_without_tracing(&plan(
        r##"{"reply":"x","actions":[{"op":"blank","color":"#ffffff"}]}"##
    )));
    assert!(loads_without_tracing(&plan(
        r##"{"reply":"x","actions":[{"op":"frame","index":0},{"op":"filter","mode":"bw"}]}"##
    )));

    // A drawn layout committed to its coordinates — but an EMPTY layout drew nothing.
    assert!(!loads_without_tracing(&plan(
        r##"{"reply":"x","actions":[{"op":"blank","color":"#ffffff"},
            {"op":"layout","lines":[{"points":[{"x":1,"y":2}]}]}]}"##
    )));
    assert!(loads_without_tracing(&plan(
        r##"{"reply":"x","actions":[{"op":"blank","color":"#ffffff"},
            {"op":"layout","lines":[]}]}"##
    )));

    // No load op ⇒ nothing new to look at; variants and an ask end the turn.
    assert!(!loads_without_tracing(&plan(
        r##"{"reply":"x","actions":[{"op":"filter","mode":"sepia"}]}"##
    )));
    assert!(!loads_without_tracing(&plan(
        r##"{"reply":"x","actions":[{"op":"blank","color":"#ffffff"}],
            "variants":[{"label":"v","actions":[{"op":"filter","mode":"bw"}]}]}"##
    )));
    assert!(!loads_without_tracing(&plan(
        r##"{"reply":"x","actions":[{"op":"blank","color":"#ffffff"}],
            "ask":{"question":"Which color?","options":[{"label":"Red"},{"label":"Blue"}]}}"##
    )));
}

File: browser/tests/opPlanAsk.test.js
// §11 interactive replies (js/llm/plan.js): the `ask` card's parse defaults, its
// option references, the malformed-card errors and askAnswerText.
import { test } from 'node:test';
import assert from 'node:assert';
import {
  parseOpPlan, validateAsk, askAnswerText, ASK_LIMITS, DEFAULT_CUSTOM_LABEL,
} from '../js/llm/plan/opPlan.js';
import { askPlan, askOf } from './helpers/opPlanRig.js';

// ── §11 interactive replies (`ask`) ─────────────────────────────────────────

test('ask: a plain single-choice card parses with its defaults filled in', () => {
  const a = askOf({ question: 'Which tint?', options: [{ label: 'Sepia' }, { label: 'B&W' }] });
  assert.equal(a.question, 'Which tint?');
  assert.equal(a.mode, 'single');              // default
  assert.equal(a.allowCustom, false);          // default
  assert.equal(a.customLabel, DEFAULT_CUSTOM_LABEL);
  assert.deepEqual(a.options.map((o) => o.label), ['Sepia', 'B&W']);
});

test('ask: absent on an ordinary plan, and on a chat-only reply', () => {
  assert.equal(parseOpPlan('{"version":1,"reply":"hi","actions":[]}').ask, null);
  assert.equal(parseOpPlan('just chatting').ask, null);
});

test('ask: multi mode, custom row, and trimmed strings', () => {
  const a = askOf({ question: '  Which ones?  ', mode: 'multi', allowCustom: true, customLabel: ' Other ',
    options: [{ label: ' A ' }, { label: 'B' }] });
  assert.equal(a.question, 'Which ones?');
  assert.equal(a.mode, 'multi');
  assert.equal(a.allowCustom, true);
  assert.equal(a.customLabel, 'Other');
  assert.equal(a.options[0].label, 'A');
});

test('ask: an option previews §2 actions — including a video frame', () => {
  const a = askOf({ question: 'Which frame?', options: [
    { label: '0:04', actions: [{ op: 'frame', index: 120 }] },
    { label: 'Sepia', actions: [{ op: 'filter', mode: 'sepia' }] },
  ] });
  assert.deepEqual(a.options[0].actions, [{ op: 'frame', index: 120 }]);
  assert.equal(a.options[1].actions[0].op, 'filter');
});

test('ask: an option names an existing image — exactly one reference kind', () => {
  const a = askOf({ question: 'Which?', options: [
    { label: 'Web', image: { url: 'https://example.com/cat.jpg' } },
    { label: 'Project', image: { projectId: 'p_12' } },
  ] });
  assert.deepEqual(a.options[0].image, { url: 'https://example.com/cat.jpg' });
  assert.deepEqual(a.options[1].image, { projectId: 'p_12' });
  // scanIndex is the extension profile's reference (§8) and stays an integer.
  assert.deepEqual(askOf({ question: 'Q', options: [{ label: 'A', image: { scanIndex: 3 } }, { label: 'B' }] })
    .options[0].image, { scanIndex: 3 });
});

test('ask: a non-http image URL is refused — a plan never points the client at data:/file:', () => {
  for (const url of ['data:image/png;base64,AAA', 'file:///etc/passwd', 'javascript:alert(1)', 'ftp://h/x.png'])
    assert.throws(() => askPlan({ question: 'Q', options: [{ label: 'A', image: { url } }, { label: 'B' }] }), /http\(s\) URL/);
});

test('ask: an option cannot be both a render and a reference', () => {
  assert.throws(() => askPlan({ question: 'Q', options: [
    { label: 'A', actions: [{ op: 'rotate', dir: 'left' }], image: { projectId: 'p1' } }, { label: 'B' },
  ] }), /both "actions" and "image"/);
});

test('ask: option counts outside 2..5 fail — a card nobody can answer is an error', () => {
  const opt = (i) => ({ label: `o${i}` });
  assert.throws(() => askPlan({ question: 'Q', options: [] }), /2\.\.5/);
  assert.throws(() => askPlan({ question: 'Q', options: [opt(1)] }), /2\.\.5/);
  assert.throws(() => askPlan({ question: 'Q', options: [1, 2, 3, 4, 5, 6].map(opt) }), /2\.\.5/);
  assert.equal(askOf({ question: 'Q', options: [1, 2, 3, 4, 5].map(opt) }).options.length, ASK_LIMITS.maxOptions);
});

test('ask: malformed cards are plan errors, not silent drops', () => {
  const two = [{ label: 'A' }, { label: 'B' }];
  assert.throws(() => askPlan({ options: two }), /"ask.question"/);
  assert.throws(() => askPlan({ question: '   ', options: two }), /"ask.question"/);
  assert.throws(() => askPlan({ question: 'x'.repeat(ASK_LIMITS.question + 1), options: two }), /longer than/);
  assert.throws(() => askPlan({ question: 'Q', options: two, mode: 'maybe' }), /"ask.mode"/);
  assert.throws(() => askPlan({ question: 'Q', options: two, allowCustom: 'yes' }), /"ask.allowCustom"/);
  assert.throws(() => askPlan({ question: 'Q', options: 'nope' }), /"ask.options" must be an array/);
  assert.throws(() => askPlan({ question: 'Q', options: two, sneaky: 1 }), /unknown field "sneaky"/);
  assert.throws(() => askPlan({ question: 'Q', options: [{ label: 'A', sneaky: 1 }, { label: 'B' }] }), /unknown field "sneaky"/);
  assert.throws(() => askPlan({ question: 'Q', options: [{ label: '' }, { label: 'B' }] }), /"label"/);
  assert.throws(() => askPlan({ question: 'Q', options: [{ label: 'x'.repeat(ASK_LIMITS.label + 1) }, { label: 'B' }] }), /longer than/);
  assert.throws(() => askPlan({ question: 'Q', options: [{ label: 'A', image: {} }, { label: 'B' }] }), /exactly one of/);
  assert.throws(() => askPlan({ question: 'Q', options: [{ label: 'A', image: { url: 'https://a/', projectId: 'p' } }, { label: 'B' }] }), /exactly one of/);
  assert.throws(() => askPlan({ question: 'Q', options: [{ label: 'A', image: { nope: 'x' } }, { label: 'B' }] }), /unknown field "nope"/);
  assert.throws(() => askPlan({ question: 'Q', options: [{ label: 'A', image: { scanIndex: -1 } }, { label: 'B' }] }), /integer >= 0/);
  assert.throws(() => askPlan({ question: 'Q', options: two, customLabel: '' }), /"ask.customLabel"/);
  assert.throws(() => askPlan({ ...{ question: 'Q', options: two } , mode: 1 }), /"ask.mode"/);
  assert.throws(() => parseOpPlan('{"version":1,"reply":"ok","ask":"hello"}'), /"ask" must be an object/);
});

test('ask: an editor-settings op in a preview drops the picture — a preview renders, never configures', () => {
  const p = askPlan({ question: 'Q', options: [
    { label: 'Dark', actions: [{ op: 'theme', mode: 'dark' }] }, { label: 'B' },
  ] });
  assert.ok(p.warnings.some((w) => /Dropped the preview for ask option 1 \("Dark"\).*not allowed inside variants/.test(w)),
    JSON.stringify(p.warnings));
  assert.strictEqual(p.ask.options[0].actions, undefined);
});

test('ask: an unknown op inside a preview is dropped with a warning, like anywhere else', () => {
  const p = askPlan({ question: 'Q', options: [
    { label: 'A', actions: [{ op: 'teleport' }, { op: 'rotate', dir: 'left' }] }, { label: 'B' },
  ] });
  assert.deepEqual(p.ask.options[0].actions.map((a) => a.op), ['rotate']);
  assert.ok(p.warnings.some((w) => /teleport/.test(w)));
});

test('ask: a plan may both edit and ask — the edits still parse', () => {
  const p = parseOpPlan(JSON.stringify({ version: 1, reply: 'cropped; now pick a tint',
    actions: [{ op: 'rotate', dir: 'right' }],
    ask: { question: 'Which tint?', options: [{ label: 'Sepia' }, { label: 'B&W' }] } }));
  assert.equal(p.actions.length, 1);
  assert.equal(p.ask.options.length, 2);
});

test('askAnswerText: the picked labels become the next user turn', () => {
  const a = askOf({ question: 'Q', mode: 'multi', options: [{ label: 'Sepia' }, { label: 'B&W' }] });
  assert.equal(askAnswerText(a, { picked: [a.options[0]] }), 'Sepia');
  assert.equal(askAnswerText(a, { picked: a.options }), 'Sepia, B&W');
  assert.equal(askAnswerText(a, { picked: ['Sepia', 'B&W'] }), 'Sepia, B&W');   // bare labels too
  assert.equal(askAnswerText(a, { picked: [] }), '');
});

test('askAnswerText: typed custom text wins, trimmed and capped', () => {
  const a = askOf({ question: 'Q', allowCustom: true, options: [{ label: 'Sepia' }, { label: 'B&W' }] });
  assert.equal(askAnswerText(a, { picked: [a.options[0]], custom: '  a warm green  ' }), 'a warm green');
  assert.equal(askAnswerText(a, { custom: 'x'.repeat(ASK_LIMITS.answer + 50) }).length, ASK_LIMITS.answer);
  assert.equal(askAnswerText(a, {}), '');
});

test('validateAsk: null in, null out (the field is optional)', () => {
  assert.equal(validateAsk(null, []), null);
  assert.equal(validateAsk(undefined, []), null);
});
